### ISSUE TITLE
ty falsely assuming arbitrary-length tuple?

### ISSUE THREAD

**Issue by andreaso** [astral-sh/ty#3100]
### Question

Mostly out of curiosity I've begun exploring running _ty_ in parallel with my existing _mypy_ checks, and when updating to to _ty_ 0.0.24 I'm getting a typing complaint that I'm not getting from _mypy_, nor was getting from _ty 0.0.11.

Since I'm not sure whatever the error is due to a bug in _ty_ or if it's due to my own limited understanding of Python typing I'm raising this as a question rather than as a bug.

Anyway, here's the complaint, following by the piece of code that the complaint is about, followed by a relevant dependent piece of code.

```
error[invalid-assignment]: Object of type `tuple[str, ...]` is not assignable to `tuple[str, str, str]`
  --> ssh_zone_handler/bind.py:32:27
   |
30 |           super().__init__(config)
31 |
32 |           self.rndc_prefix: Final[tuple[str, str, str]] = self.sudo_prefix + (
   |  ___________________________---------------------------___^
   | |                           |
   | |                           Declared type
33 | |             "/usr/sbin/rndc",
34 | |         )
   | |_________^ Incompatible value of type `tuple[str, ...]`
35 |
36 |       def __lookup(self, zone: str, failure: str) -> str | None:
   |
info: rule `invalid-assignment` is enabled by default
```
```python
class BindCommand(SshZoneCommand):
    """Runs the actual commands, for BIND"""

    def __init__(self, config: ZoneHandlerConf) -> None:
        super().__init__(config)

        self.rndc_prefix: Final[tuple[str, str, str]] = self.sudo_prefix + (
            "/usr/sbin/rndc",
        )
```
```python
class SshZoneCommand(SshZoneHandler):
    """Command class to runs the actual commands"""

    def __init__(self, config: ZoneHandlerConf) -> None:
        super().__init__(config)

        self.sudo_prefix: Final[tuple[str, str]] = (
            "/usr/bin/sudo",
            f"--user={self.service_user}",
        )
```

So, the way I see it `self.sudo_prefix` is already defined without complaint as a fixed-length `tuple[str, str]` and with `("/usr/sbin/rndc", )` explicitly containing one string I would think it clear that the result would be a fixed length fixed-length `tuple[str, str, str]`? Yet _ty_ appear to see it as an arbitrary-length `tuple[str, ...]`?

Bug, or something I simply have misunderstood?

### Version

ty 0.0.24

**Comment by K-Mirembe-Mercy** [comment]
Hi! I think this is due to how type inference works when concatenating tuples. Even though both tuples have fixed lengths, many type checkers fall back to "tuple[str, ...]" when using "+", because they don't always preserve the exact tuple size after concatenation.

So "self.sudo_prefix" is correctly inferred as "tuple[str, str]", but when you do:

self.sudo_prefix + ("/usr/sbin/rndc",)

the result is treated as a variable-length tuple of strings rather than a fixed-length "tuple[str, str, str]".

This behavior is similar in some cases with other type checkers as well, so it might be more of a limitation of tuple concatenation inference rather than a bug in "ty".

One workaround could be to explicitly type the result or construct the tuple directly instead of using "+".

Hope that helps

**Comment by AlexWaygood** [comment]
Thanks for the report! Please see #1504

**Comment by andreaso** [comment]
@K-Mirembe-Mercy, @AlexWaygood: Thanks!
